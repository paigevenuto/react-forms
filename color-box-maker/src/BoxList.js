import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import _ from "lodash";
import { v4 as uuid } from "uuid";

function BoxList() {
  const [boxes, setBoxes] = useState([]);

  const delBox = (evt) => {
    console.log(evt);
    const boxesClone = boxes.filter(
      (b) => b.id !== evt.target.parentElement.id
    );
    setBoxes(boxesClone);
  };

  const addBox = (formData) => {
    let { backgroundColor, width, height } = formData;
    width = width + "px";
    height = height + "px";
    const id = uuid();
    const boxesClone = _.cloneDeep(boxes);
    boxesClone.push({ backgroundColor, width, height, id });
    setBoxes(boxesClone);
  };

  return (
    <div className="BoxList" data-testid="BoxList">
      <NewBoxForm addBox={addBox} />
      {boxes.map((b) => (
        <Box
          backgroundColor={b.backgroundColor}
          width={b.width}
          height={b.height}
          delBox={delBox}
          id={b.id}
          key={b.id}
        />
      ))}
    </div>
  );
}

export default BoxList;
