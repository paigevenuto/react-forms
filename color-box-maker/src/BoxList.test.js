import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// smoke test
it("renders without crashing", function () {
  render(<BoxList />);
});

// snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("creates a red 300x250 box", function () {
  const { getByLabelText, queryByText, getAllByTestId } = render(<BoxList />);

  // select the "red" "300" and "250" options
  const widthField = getByLabelText("Box Width:", {
    selector: "input",
  });
  const heightField = getByLabelText("Box Height:", {
    selector: "input",
  });
  const backgroundColorField = getByLabelText("Box Color:", {
    selector: "input",
  });
  const submitBtn = queryByText("Add!");
  fireEvent.change(backgroundColorField, { target: { value: "red" } });
  fireEvent.change(widthField, { target: { value: "300" } });
  fireEvent.change(heightField, { target: { value: "250" } });

  expect(widthField.value).toBe("300");
  expect(heightField.value).toBe("250");
  expect(backgroundColorField.value).toBe("red");
  fireEvent.click(submitBtn);
  expect(widthField.value).toBe("");
  expect(heightField.value).toBe("");
  expect(backgroundColorField.value).toBe("");

  const boxes = getAllByTestId("Box");
  expect(boxes[0]).toHaveStyle("background-color: red;");
  expect(boxes[0]).toHaveStyle("width: 300px;");
  expect(boxes[0]).toHaveStyle("height: 250px;");
});
