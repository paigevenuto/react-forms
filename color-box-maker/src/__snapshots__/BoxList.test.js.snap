// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
    >
      <label
        for="backgroundColor"
      >
        Box Color:
      </label>
      <input
        list="backgroundColors"
        name="backgroundColor"
        required=""
        value=""
      />
      <datalist
        id="backgroundColors"
      >
        <option
          value="red"
        />
        <option
          value="blue"
        />
        <option
          value="green"
        />
        <option
          value="pink"
        />
        <option
          value="teal"
        />
        <option
          value="purple"
        />
        <option
          value="orange"
        />
        <option
          value="yellow"
        />
        <option
          value="magenta"
        />
      </datalist>
      <label
        for="width"
      >
        Box Width:
      </label>
      <input
        name="width"
        required=""
        type="number"
        value=""
      />
      <label
        for="height"
      >
        Box Height:
      </label>
      <input
        name="height"
        required=""
        type="number"
        value=""
      />
      <button>
        Add!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
